### ISSUE TITLE
Rethink `SharedObservation` as an auto-updating cache

### ISSUE THREAD

**Issue by groue** [groue/GRDB.swift#1093]
💭 It is uncanny how much [SharedObservation](https://github.com/groue/GRDB.swift/pull/1088) (not shipped on the master branch yet) looks like a cache.

Since it is always able to notify its [latest fetched value](https://github.com/groue/GRDB.swift/blob/92e4c8f785beb3d951a72fab1f949ac26bb56591/GRDB/ValueObservation/SharedValueObservation.swift#L127) to new subscribers, it becomes natural to just expose it publicly:

```swift
let sharedObservation = ValueObservation.tracking { ... }.shared(in: ...)

if let value = sharedObservation.value {
    // Already cached
} else {
    // Not cached yet, or got a database error recently
}
```

The problem is that now `SharedObservation` looks like a wrong name, as well as the `start` method.

It's time to look at existing alternatives, grab inspiration, and repackage `SharedObservation`.

For example, [NSFetchedResultsController](https://developer.apple.com/documentation/coredata/nsfetchedresultscontroller) supports:
- change notification (so do we)
- access to latest fetched value (so can we)
- request change (that's an interesting feature)
- diffing ([we don't care](https://github.com/groue/GRDB.swift/issues/1086#issuecomment-949491402))

Back to the drawing board. `SharedObservation` is unlikely to ship as is.

Sorry @tcwalther, this might delay the next version for a while 😅

_Originally posted by @groue in https://github.com/groue/GRDB.swift/pull/1088#issuecomment-952673109_

**Comment by groue** [comment]
After tickling with the idea for a while, I discovered that it was difficult to offer a consistent user experience regarding the _latest fetched value_.

The difficulty lies in the default extent of the shared value observation, `.whileObserved`:

```swift
let sharedObservation = observation.shared(in: dbQueue, extent: .whileObserved)
```

Before the observation has started, the latest database value is not known yet.

When observation starts, it becomes known.

When all subscriptions end, or when the observation fails due to a database error, the database observation stops. The latest database value becomes unknown again. Only a new subscription can put the latest database value back to a known state.

This means that "the latest database value" _depends_ on the number of existing subscriptions. It is not an intrinsic property of a `SharedValueObservation` instance.

Conclusion: `SharedValueObservation` _is not_ a cache.

**Comment by tcwalther** [comment]
@groue thank you so much for putting so much thought into it. I admire the thoroughness with which you approach this project.

I'm sure you already though about this, but how about nilling the value or throwing an exception on value access when no observation is running? It might be perfectly acceptable to require developers to make sure at least one observation is running when accessing this value.

**Comment by groue** [comment]
> but how about nilling the value or throwing an exception on value access when no observation is running?

Yes, this is a possibility. Nil or error would means "I don't know yet", or "I used to know, but observation has stopped because an error has occurred", or "I used to know, but observation has stopped because there is no active subscription".

> It might be perfectly acceptable to require developers to make sure at least one observation is running when accessing this value.

I think it is not. It is _possible_ to describe what nil or error would mean, but this description is very complex, and does not explain how to deal with errors, or how to deal with multi-threading, and this is all very difficult in practice.

If I would ship an api that _looks like_ you're very close to an auto-updating cache, then users would write awful code full of bugs in order to complete the picture. Then I would have to answer GItHub issues and questions, and explain how to do it properly. And I will regret having shipped such a lousy api, because it frustrates everybody, users and me.

On the other side, I think it's possible, in userland, to write an "auto-updating cache" that _wraps_ a ValueObservation, shared or not. This is not trivial either, but this is now the _full responsibility of the user_. This is the ideal situation. The user is not frustrated because she can use reliable apis, and reach her goal with some elbow grease. And I'm not frustrated because I'm not even aware.
